Extend description of PAML purpose

rpgoldman opened the pull request:
Add an enumerated list of tasks that PAML should support.  Intended to provide a path to identifying requirements and assessing our success in meeting those requirements.

rpgoldman committed:
Add more details to the purpose statement.

Add an itemized list of subtasks from which to extract requirements.

jakebeal commented on purpose.tex:
```diff
@@ -5,13 +5,34 @@ \section{Purpose}
+  \begin{description}
+  \item[Instantiation Proper] We must be able to create some persistent data structure that represents the protocol instance.
+  \item[Derivation] We must be able to represent the relationship between the protocol instance and the recipe from which it is derived.
+  \item[Modification] We must be able to record modifications made to the original recipe, for example, to enable it to be executed at a lab with different equipment from the lab at which it was originated, to enable the protocol to be scaled up or scaled down, etc.
```
I'm not sure this is something that we want to record under Instantiation --- I think this is a matter of making a derived work and recording the relationship.  If we want support execution of modified protocols, that goes against the decision to disallow dynamical patching of protocols that is recorded in https://github.com/SD2E/paml/issues/15

A use pattern that I think may end up being common is to have a protocol be "too strict" to be instantiated in a lab, and thus need modification. Rather than the modification being a "patch", however, it would be a generalization that allows it to be instantiated in both prior labs and the new labs, which would be released as an updated version of the existing protocol. The instantiation would then not be a modification, but an instantiation of a later version than first intended.

rpgoldman replied:
I have captured that in a new version that I will push momentarily, incorporating your example.

One issue about instantiation is what we think instantiation should be. Is instantiation very much like making a new version of the protocol (so relatively complete)? Or do we have in mind something where the instance is a different sort of object, more lightweight, that simply records a link between a block of parameters (OPIL, presumably), the protocol (PAML), and possibly some additional metadata and annotation?  Either seems possible, but we should probably state a rationale for choosing one or the other.

jakebeal replied:
@rpgoldman I expect that we'll have both: as you note, the lightweight is OPIL, the heavy-weight is an execution trace for PAML.

jakebeal commented on purpose.tex:
```diff
@@ -5,13 +5,34 @@ \section{Purpose}
+\item[Planning and scheduling] Lab resources are valuable and scarce, and some organizations will want to be able to optimize their use. To do so, PAML should support (at least) extraction of resource requirements from activities in the protocol, and estimated durations.  Note that \emph{what} resource requirements and duration estimates are required will likely be a function of both the protocol ...
+\item[Verification] Authoring a formal protocol in PAML requires substantial mental effort, and the usefulness of the corresponding artifacts can be compromised if they are ill-formed or erroneous.  Supporting protocol authors in achieving correctness is an important goal. Some of this will fall onto the shoulders of those building PAML support tools, however the specification itself must provide ...
+\item[Mapping protocols to labs] PAML aims to support replication in part by providing automated support for mapping protocols developed at one lab onto another.  We suspect that for some time to come this will be at best a semi-automated process, and like verification, this task will fall heavily on those who provide tool support for PAML users.  However, also as for verification, the PAML speci ...
+\item[``Derived works"] As with most complex formal artifacts (programs, spreadsheets, etc.), we confidently expect that few users will create new protocols from an entirely blank slate.  Instead, most will take an existing protocol, copy it, and edit the copy.  PAML must support this mode of operation.  Key issues here are which component structures can be incorporated by reference, and which mu ...
```
I think it would be good to also mention ongoing development and versioning of a protocol.

rpgoldman replied:
Done.  One issue we might want to think about is when we are refining and existing protocol to make a new version, and when we are creating a derived protocol.

This is not going to be something that we can provide a hard, crisp answer to, but it is something about which we might wish to provide guidance.

jakebeal reviewed (changes requested):
Looking pretty good; I've added two suggestions

rpgoldman committed:
Incorporate improvements from @jakebeal.

1. Refine the discussion of the modification task and move it to top
level, rather than being part of "Instantiation," as
before. Incorporate an example from Jake.

2. Add a "Protocol Maintenance" task.

3. Revise title of "Derived works" task to be an action, like the
rest, instead of a noun (not from Jake).

rpgoldman commented:
@jakebeal I suspect we should add a "requirements" section or something like that, and this purpose material should be boiled down to a simple list, with cross references into the requirements chapter. Or "Requirements and design rationale."

jakebeal commented:
Agreed. In the SBOL specs, we've had an "Overview" section which provides a high-level picture and design rationale.

jakebeal commented:
Why don't you merge these changes in for now, and then set up that next modification as a new branch?

rpgoldman committed:
Add "Execution" task description.

rpgoldman commented:
> Why don't you merge these changes in for now, and then set up that next modification as a new branch?

I will do that. Would you please just have a look at the "Execution" item I added. If you are OK with that, I will merge.

jakebeal commented:
Looks good to me; please merge.
